docs(spec): Add specification for deprecations (#14896)
influxdata/telegraf: docs/specs/tsd-001-deprecation.md

plugin features including a timeline and minimal time-frames. Additionally, the
specification defines a framework to annotate deprecations in the code and
inform users about such deprecations.

## User experience

In the deprecation phase a warning will be shown at Telegraf startup with the
following content

```text
Plugin "inputs.logparser" deprecated since version 1.15.0 and will be removed in 2.0.0: use 'inputs.tail' with 'grok' data format instead
```

Similar warnings will be shown when removing plugin options or option values.
This provides users with time to replace the deprecated plugin in their
configuration file.

After the shown release (`v2.0.0` in this case) the warning will be promoted
to an error preventing Telegraf from starting. The user now has to adapt the
configuration file to start Telegraf.

## Deprecation Process

After reaching an agreement in the issue, you can start the deprecation process
by following the steps below.

### File issue

In the filed issue you should outline which plugin, plugin option or feature
you want to deprecate and *why*! Determine in which version the plugin should

> be removed.

Reviewer: Can we add a new line below here.

> predate the removal by at least one and a half year. In current release terms

Reviewer 2: ```suggestion
predate the removal by at least one and a half years. In current release terms
```

Consider moving the plugin to the Influx community repository as an external

> plugin if there is no straight-forward replacement.

Reviewer: Would we actually deprecate a plugin and remove it to an external plugin if there was no replacement? Later on we talk about not deprecating anything unless there is actually a replacement ;) so these seems to conflict.

my suggestion is to remove this paragraph.

> requiring a pro-longed time between deprecation and removal to ensure a smooth

Reviewer 2: ```suggestion
requiring a prolonged time between deprecation and removal to ensure a smooth
```

Try to reach an agreement in the issue before continuing and get a sign off
from the maintainers!

### Submit deprecation pull-request

Send a pull request adding deprecation information to the code and update the
plugin's `README.md` file. Depending on what you want to deprecate this
comprises different locations and steps as detailed below.

Once the deprecation pull-request is merged and Telegraf is released, we have
to wait for the targeted Telegraf version for actually removing the code.

You should consider writing a blog-post to make users aware of the scheduled

> removal of plugins, plugin options or option values.

Reviewer: You? Or the maintainers? Up to this point it reads like an external user would do all of this, but I would not expect an external user to write the blog post for this.

I also don't think we would do a blog post for every removal. Instead we would do a big post maybe one release before? or at the release we remove things in?

Author: This is a copy and past from the original document. So should I remove this paragraph?

Reviewer: Yeah let's remove

#### Deprecating a plugin

When deprecating a plugin you need to add an entry to the `deprecation.go` file
in the respective plugin category with the following format

```golang
    "<plugin name>": {
        Since:     "<x.y.z format version of the next minor release>",
        RemovalIn: "<x.y.z format version of the plugin removal>",
        Notice:    "<user-facing hint e.g. on replacements>",
    },
```

If you for example want to remove the `inputs.logparser` plugin you should add

```golang
    "logparser": {
        Since:     "1.15.0",
        RemovalIn: "1.40.0"
        Notice:    "use 'inputs.tail' with 'grok' data format instead",
    },
```

> to `plugins/inputs.deprecations.go`. By doing this, Telegraf will show a

Reviewer 3: ```suggestion
to `plugins/inputs/deprecations.go`. By doing this, Telegraf will show a
```

deprecation warning to the user starting from version `1.15.0` including the
`Notice` you provided. The plugin can then be remove in version `1.40.0`.

Additionally, you should update the plugin's `README.md` stating the plugin is

> deprecated at the top (e.g. ***DEPRECATED***).

Reviewer: Can you add an example, as I don't think you want the user to just put `***DEPRECATED***`, you want that + the notice to help guide the user.

Author: Again a copy and paste from the original document. For the now removed plugins that was exactly what was done. Should I change to what is done in `inputs.logparser` 

```
**Deprecated in Telegraf 1.15: Please use the [tail][] plugin along with the
[`grok` data format][grok parser]**
```
instead?

Reviewer: you should feel free to update the specs as you put them up here

> Should I change to what is done in inputs.logparser

Yes I like your proposed change. We should ensure we guide the user with the call to action of what to move to next.

Reviewer 3: I do like more how it is done in the aerospike plugin: https://github.com/influxdata/telegraf/blob/a8551659bc8ce34d13d37c58da31ebcb3f469d7e/plugins/inputs/aerospike/README.md?plain=1#L3-L5

I used it the same way to deprecate the ifname processor: https://github.com/influxdata/telegraf/pull/14899/files#diff-ad714a45b893f49a07ddd5132428edea0bd32db5650233832a200705738c0401R3-R4

Before the removal of deprecated plugins #13376, most of them were using this format:
```markdown
**Deprecated in version 1.30**: Please use [snmp_lookup][] plugin.
```

So should it be _version_ or _Telegraf_ in that section? To me te first looks better.

Author: Both versions would be possible according to this spec. I don't see why we would need to have `version of Telegraf` as all of this is version of Telegraf as we don't have any other version...

Reviewer 3: Oh it was meant like a suggestion then? 👍

I didn't mean to put `version of Telegraf`, but `Deprecated in version 1.30` instead of `Deprecated in Telegraf 1.30`.

#### Deprecating an option

> To deprecate a plugin open, remove the option from the `sample.conf` file and

Reviewer: ```suggestion
To deprecate a plugin option, remove the option from the `sample.conf` file and
```

            Since:     "1.23.1",
            RemovalIn: "1.40.0",
            Notice:    "use 'v1' instead",
        })
    }
    ...
    return nil
}
```

This will show a warning if the deprecated `v1` value is used for the `mode`
setting in `inputs.example` with a user-facing notice.

### Submit pull-request for removing code

Once the plugin, plugin option or option-value is deprecated, we have to wait
for the `RemovedIn` release to remove the code. In the examples above, this
would be version `1.40.0`. After all scheduled bugfix-releases are done, with
`1.40.0` being the next release, you can create a pull-request to actually
remove the deprecated code.

Please make sure, you remove the plugin, plugin option or option value and the
code referencing those. This might also comprise the `all` files of your plugin
category, test-cases including those of other plugins, README files or other
documentation. For removed plugins, please keep the deprecation info in
`deprecations.go` so users can find a reference when switching from a really
old version.

Make sure you add an `Important Changes` sections to the `CHANGELOG.md` file
describing the removal with a reference to your PR.

> ### Time frames and considerations

Reviewer: I feel like this section should go towards the top or with the section with user-experience, not at the very end.

When deprecating parts of Telegraf, it is important to provide users with enough
time to migrate to alternative solutions before actually removing those parts.

In general, plugins, plugin options or option values should only be deprecated
if a suitable alternative exists! In those cases, the deprecations should
predate the removal by at least one and a half year. In current release terms
this corresponds to six minor-versions. However, there might be circumstances
requiring a pro-longed time between deprecation and removal to ensure a smooth
transition for users.

> ## Deprecation Framework

Reviewer: This whole section has pretty much already been explained, right? I'm not sure what this adds?

Author: It specifies that there should be a framework doing the warning and then error-and-exit. Should I remove this section?

Reviewer: Yeah, I would, otherwise this feels out of place.

to wait for the targeted Telegraf version for actually removing the code.

#### Deprecating a plugin

When deprecating a plugin you need to add an entry to the `deprecation.go` file
in the respective plugin category with the following format

```golang
    "<plugin name>": {
        Since:     "<x.y.z format version of the next minor release>",
        RemovalIn: "<x.y.z format version of the plugin removal>",
        Notice:    "<user-facing hint e.g. on replacements>",
    },
```

If you for example want to remove the `inputs.logparser` plugin you should add

```golang
    "logparser": {
        Since:     "1.15.0",
        RemovalIn: "1.40.0"
        Notice:    "use 'inputs.tail' with 'grok' data format instead",
    },
```

to `plugins/inputs/deprecations.go`. By doing this, Telegraf will show a
deprecation warning to the user starting from version `1.15.0` including the
`Notice` you provided. The plugin can then be remove in version `1.40.0`.

Additionally, you should update the plugin's `README.md` stating the plugin is
deprecated by adding a `***DEPRECATED***` tag next to the plugin title at the
top *and* by adding a small section mentioning since when the plugin is

> deprecated, when it will be removed and a hint to alternatives like

Reviewer: Agreed we should remove this bit. The docs team is working on updating the docs based on our in-repo docs and I wouldn't want the titles updated like this either.
